src/aws-route53/split-horizon-dns.ts

```diff
@@ -0,0 +1,165 @@
+import * as cdk from 'aws-cdk-lib';
+import * as acm from 'aws-cdk-lib/aws-certificatemanager';
+import * as ec2 from 'aws-cdk-lib/aws-ec2';
+import * as route53 from 'aws-cdk-lib/aws-route53';
+import { Construct } from 'constructs';
+
+const isAliasRecordTarget = (value: route53.IAliasRecordTarget | Array<string>): value is route53.IAliasRecordTarget => {
+  return !('length' in value);
+};
+
+type ARecordArray = Array<route53.ARecord>;
+
+export interface AliasTarget {
+  readonly target: route53.IAliasRecordTarget | Array<string>;
+  readonly private?: boolean;
+  readonly public?: boolean;
+  readonly ttl?: cdk.Duration;
+}
+
+/**
+ * @param zoneName The DNS name of the zone to create
+ * @param existingPublicZone An existing public zone to use instead of creating a new one
+ * @param existingPrivateZone An existing private zone to use instead of creating a new one
+ * @param disallowPrivateZone Override the default behavior of creating a private zone. Will also block adding private records.
+ * @param includeCertificate Whether to create an ACM certificate for the zone
+ * @param certAlternateNames Alternate names to include in the certificate
+ * @param privateZoneVpcs VPCs to associate with the private zone
+ * @param targets Targets to create A records for
+ */
+export interface ISplitHorizonDnsProps {
+  readonly zoneName: string;
+  readonly recordName?: string;
+  readonly existingPublicZone?: route53.IHostedZone;
+  readonly existingPrivateZone?: route53.IHostedZone;
+  readonly disallowPrivateZone?: boolean;
+  readonly certAlternateNames?: Array<string>;
+  readonly privateZoneVpcs?: Array<ec2.Vpc>;
+  readonly targets: Array<AliasTarget>;
+  readonly includeCertificate?: boolean;
+}
+
+export interface ISplitHorizonDns {
+  publicZone: route53.IHostedZone;
+  privateZone?: route53.IHostedZone;
+  records: Array<ARecordArray>;
+}
+
+interface createHostedZoneProps {
+  zoneName: string;
+  vpcs?: Array<ec2.IVpc>;
+}
+
+interface createCertificateProps {
+
+  domainName: string;
+  subjectAlternativeNames?: Array<string>;
+}
+/**
+ * Creates a public and private zone for a given domain name, and creates A records for the given targets.
+ * @property publicZone The public zone created
+ * @property privateZone The private zone created
+ * @property records The A records created
+ */
+export class SplitHorizonDns extends Construct implements ISplitHorizonDns {
+  public publicZone: route53.IHostedZone;
+
+  public privateZone?: route53.IHostedZone;
+
+  public records: Array<ARecordArray>;
+
+  public certificate?: acm.ICertificate;
+
+  constructor(scope: Construct, id: string, private props: ISplitHorizonDnsProps) {
```

Comment:
please make this protected so sub-classes can access this info

Reply:
I'm not sure I follow why a protected constructor is needed here?